parquet-encoding/src/main/java/org/apache/parquet/bytes/CapacityByteBufferOutputStream.java

```diff
@@ -190,11 +197,32 @@ public void write(byte b[], int off, int len) {
    * @exception  IOException  if an I/O error occurs.
    */
   public void writeTo(OutputStream out) throws IOException {
+    ByteBuffer slab;
     for (int i = 0; i < slabs.size() - 1; i++) {
-      final byte[] slab = slabs.get(i);
-      out.write(slab);
+      slab = slabs.get(i);
+      if (slab.hasArray()) {
+        out.write(slab.array(), slab.arrayOffset(), slab.position());
+      } else {
+        // The OutputStream interface only takes a byte[], unfortunately this means that a ByteBuffer
+        // not backed by a byte array must be copied to fulfil this interface
```

Comment:
Looking at the change that introduced the Channel use, it was removing what was previously a loop calling the get() method on the ByteBuffer and writing each byte individually. The docs on WritableByteChannel are concerned mostly with concurrency, which I'm pretty sure isn't an issue here. I think it might make sense to refactor the other one to use this method. I can try to factor out the code somewhere between the two.